RecoPixelVertexing/PixelVertexFinding/plugins/gpuVertexFinderImpl.h

```diff
@@ -0,0 +1,182 @@
+#include "HeterogeneousCore/CUDAUtilities/interface/cudaCheck.h"
+
+#include "gpuClusterTracksByDensity.h"
+#include "gpuClusterTracksDBSCAN.h"
+#include "gpuClusterTracksIterative.h"
+#include "gpuFitVertices.h"
+#include "gpuSortByPt2.h"
+#include "gpuSplitVertices.h"
+
+#undef PIXVERTEX_DEBUG_PRODUCE
+
+namespace gpuVertexFinder {
+
+  __global__ void loadTracks(TkSoA const* ptracks, ZVertexSoA* soa, WorkSpace* pws, float ptMin) {
+    assert(ptracks);
+    assert(soa);
+    auto const& tracks = *ptracks;
+    auto const& fit = tracks.stateAtBS;
+    auto const* quality = tracks.qualityData();
+
+    auto first = blockIdx.x * blockDim.x + threadIdx.x;
+    for (int idx = first, nt = TkSoA::stride(); idx < nt; idx += gridDim.x * blockDim.x) {
+      auto nHits = tracks.nHits(idx);
+      if (nHits == 0)
+        break;  // this is a guard: maybe we need to move to nTracks...
+
+      // initialize soa...
+      soa->idv[idx] = -1;
+
+      if (nHits < 4)
+        continue;  // no triplets
+      if (quality[idx] != pixelTrack::Quality::loose)
+        continue;
+
+      auto pt = tracks.pt(idx);
+
+      if (pt < ptMin)
+        continue;
+
+      auto& data = *pws;
+      auto it = atomicAdd(&data.ntrks, 1);
+      data.itrk[it] = idx;
+      data.zt[it] = tracks.zip(idx);
+      data.ezt2[it] = fit.covariance(idx)(14);
+      data.ptt2[it] = pt * pt;
+    }
+  }
+
+// #define THREE_KERNELS
+#ifndef THREE_KERNELS
+  __global__ void vertexFinderOneKernel(gpuVertexFinder::ZVertices* pdata,
+                                        gpuVertexFinder::WorkSpace* pws,
+                                        int minT,      // min number of neighbours to be "seed"
+                                        float eps,     // max absolute distance to cluster
+                                        float errmax,  // max error to be "seed"
+                                        float chi2max  // max normalized distance to cluster,
+  ) {
+    clusterTracksByDensity(pdata, pws, minT, eps, errmax, chi2max);
+    __syncthreads();
+    fitVertices(pdata, pws, 50.);
+    __syncthreads();
+    splitVertices(pdata, pws, 9.f);
+    __syncthreads();
+    fitVertices(pdata, pws, 5000.);
```

Comment:
50, 9, 5000 are repeated multiple times; this looks like a place to define a common constant or a configurable

Reply:
@VinInn can you suggest names for these ?

Reply:
cut0, cut1, cut2?

Reply:
or 
maxChi2ForFirstFit
maxChi2ForSplit
maxChi2ForSecondFit  (or maxChi2ForFinalFit)

Reply:
They are repeated as the corresponding function calls.
IN principle we can remove all the alternative schedules and keep only the single block version.

Reply:
OK, maybe `chi2maxInitialFit`, `chi2maxSplitting`, `chi2maxFinalFit` ?

I don't know what they do, so I'm kind of making the names up...

Reply:
for the fit is an outlier rejection cut: if a track contributes for more than chi2max is rejected.
in the splitting is more a chi2minForSpilling i.e. only vertices with chi2/Ndif > chi2Min will be split